src/lavinmq/mqtt/retain_store.cr

```diff
@@ -0,0 +1,173 @@
+            next
+          end
+          index.insert(topic, msg_file_name)
+          Log.debug { "restored #{topic}" }
+          msg_count += 1
+        end
+
+        # TODO: Device what's the truth: index file or msgs file. Mybe drop the index file and rebuild
+        # index from msg files?
+        unless msg_file_segments.empty?
+          Log.warn { "unreferenced messages will be deleted: #{msg_file_segments.join(",")}" }
+          msg_file_segments.each do |file_name|
+            File.delete? File.join(dir, file_name)
+          end
+        end
+        Log.info { "restoring index done, msg_count = #{msg_count}" }
+      end
+
+      def retain(topic : String, body_io : ::IO, size : UInt64) : Nil
+        @lock.synchronize do
+          Log.debug { "retain topic=#{topic} body.bytesize=#{size}" }
+          # An empty message with retain flag means clear the topic from retained messages
+          if size.zero?
+            delete_from_index(topic)
+            return
+          end
+
+          unless msg_file_name = @index[topic]?
+            msg_file_name = make_file_name(topic)
+            add_to_index(topic, msg_file_name)
+          end
+
+          tmp_file = File.join(@dir, "#{msg_file_name}.tmp")
+          File.open(tmp_file, "w+") do |f|
+            f.sync = true
+            ::IO.copy(body_io, f)
+          end
+          final_file_path = File.join(@dir, msg_file_name)
+          File.rename(tmp_file, final_file_path)
+          @files.delete(final_file_path)
+          @files[final_file_path] = File.new(final_file_path, "r+")
+          @replicator.replace_file(final_file_path)
+        ensure
+          FileUtils.rm_rf tmp_file unless tmp_file.nil?
+        end
+      end
+
+      private def write_index
+        tmp_file = File.join(@dir, "#{INDEX_FILE_NAME}.next")
+        File.open(tmp_file, "w+") do |f|
+          @index.each do |topic, _filename|
+            f.puts topic
+          end
+        end
+        File.rename tmp_file, @index_file_name
+        @replicator.replace_file(@index_file_name)
+      ensure
+        FileUtils.rm_rf tmp_file unless tmp_file.nil?
+      end
+
+      private def add_to_index(topic : String, file_name : String) : Nil
+        @index.insert topic, file_name
+        @index_file.puts topic
+        @index_file.flush
+        bytes = Bytes.new(topic.bytesize + 1)
+        bytes.copy_from(topic.to_slice)
+        bytes[-1] = 10u8
+        @replicator.append(@index_file_name, bytes)
+      end
+
+      private def delete_from_index(topic : String) : Nil
+        if file_name = @index.delete topic
+          Log.trace { "deleted '#{topic}' from index, deleting file #{file_name}" }
+          if file = @files[file_name]
+            @files.delete(file)
+            file.close
+            file.delete
+          end
+          @replicator.delete_file(File.join(@dir, file_name))
+        end
+      end
+
+      def each(subscription : String, &block : String, Bytes -> Nil) : Nil
```

Comment:
in the only location we use this method we convert the body to a String, which will raise if there is binary data that can't be parsed as utf8, but also, why don't we just return the File/IO object? 
```suggestion
      def each(subscription : String, &block : String, IO -> Nil) : Nil
```